commit d1044f39efd9240a99762f108da5728845464a82
Author: Jens Axboe <axboe@kernel.dk>
Date:   Tue Jan 17 14:22:24 2017 -0800

    iwlwifi: fix kernel crash when unregistering thermal zone
    
    
    [ Upstream commit 92549cdc288f47f3a98cf80ac5890c91f5876a06 ]
    
    A recent firmware change seems to have enabled thermal zones on the
    iwlwifi driver. Unfortunately, my device fails when registering the
    thermal zone. This doesn't stop the driver from attempting to unregister
    the thermal zone at unload time, triggering a NULL pointer deference in
    strlen() off the thermal_zone_device_unregister() path.
    
    Don't unregister if name is NULL, for that case we failed registering.
    Do the same for the cooling zone.
    
    Signed-off-by: Jens Axboe <axboe@fb.com>
    Signed-off-by: Kalle Valo <kvalo@codeaurora.org>
    Signed-off-by: Sasha Levin <alexander.levin@verizon.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/drivers/net/wireless/intel/iwlwifi/mvm/tt.c b/drivers/net/wireless/intel/iwlwifi/mvm/tt.c
index 63a051b..bec7d9c 100644
--- a/drivers/net/wireless/intel/iwlwifi/mvm/tt.c
+++ b/drivers/net/wireless/intel/iwlwifi/mvm/tt.c
@@ -843,8 +843,10 @@ static void iwl_mvm_thermal_zone_unregister(struct iwl_mvm *mvm)
 		return;
 
 	IWL_DEBUG_TEMP(mvm, "Thermal zone device unregister\n");
-	thermal_zone_device_unregister(mvm->tz_device.tzone);
-	mvm->tz_device.tzone = NULL;
+	if (mvm->tz_device.tzone) {
+		thermal_zone_device_unregister(mvm->tz_device.tzone);
+		mvm->tz_device.tzone = NULL;
+	}
 }
 
 static void iwl_mvm_cooling_device_unregister(struct iwl_mvm *mvm)
@@ -853,8 +855,10 @@ static void iwl_mvm_cooling_device_unregister(struct iwl_mvm *mvm)
 		return;
 
 	IWL_DEBUG_TEMP(mvm, "Cooling device unregister\n");
-	thermal_cooling_device_unregister(mvm->cooling_dev.cdev);
-	mvm->cooling_dev.cdev = NULL;
+	if (mvm->cooling_dev.cdev) {
+		thermal_cooling_device_unregister(mvm->cooling_dev.cdev);
+		mvm->cooling_dev.cdev = NULL;
+	}
 }
 #endif /* CONFIG_THERMAL */
 
